src/types/web-resource.ts

```diff
@@ -0,0 +1,155 @@
+import * as TypeUtils from '../type-utils';
+import { ContentOptions, getStorageAdapter } from '../storage-adapters/index';
+
+/**
+ * This is how WebResources are stored on the DB
+ */
+export type WebResourceRef = {
+	filename: string;
+	href: string;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+};
+
+/**
+ * This is how WebResources are defined on creation
+ */
+export type WebResourceInput = {
+	filename: string;
+	data: Buffer;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+	storage: string;
+};
+
+export const types = {
+	postgres: 'JSONB',
+	mysql: 'JSON',
+	websql: 'TEXT',
+	odata: {
+		name: 'Self.WebResource',
+		complexType: `\
+<ComplexType Name="WebResource">
+	<Property Name="filename" Nullable="false" Type="Edm.String"/>\
+	<Property Name="href" Nullable="false" Type="Edm.String"/>\
+	<Property Name="contentType" Nullable="true" Type="Edm.String"/>\
+	<Property Name="contentDisposition" Nullable="true" Type="Edm.String"/>\
+	<Property Name="size" Nullable="true" Type="Edm.Int64"/>\
+</ComplexType>`,
+	},
+};
+
+export const nativeProperties = {
+	has: {
+		Filename: (from: any) => [
+			'ExtractJSONPathAsText',
+			from,
+			['TextArray', ['EmbeddedText', 'filename']],
+		],
+		HRef: (from: any) => [
+			'ExtractJSONPathAsText',
+			from,
+			['TextArray', ['EmbeddedText', 'href']],
+		],
+		'Content Type': (from: any) => [
+			'ExtractJSONPathAsText',
+			from,
+			['TextArray', ['EmbeddedText', 'contentType']],
+		],
+		'Content Disposition': (from: any) => [
+			'ExtractJSONPathAsText',
+			from,
+			['TextArray', ['EmbeddedText', 'contentDisposition']],
+		],
+		Size: (from: any) => [
+			'ExtractJSONPathAsText',
+			from,
+			['TextArray', ['EmbeddedText', 'size']],
+		],
```

Comment:
What do they return on empty JSON values for those keys, is it undefined?

Reply:
@fisehara I'm not sure how to validate this. I've added constraints ( "necessities" ) on a test model to be able to use them, for example:
`	Necessity: each student that has a picture, has a picture that has a Content Disposition (Type) that is equal to "inline"`

but I'm not sure how to directly validate accessing them. 
I'll look for an example of how this can be tested

Reply:
@joshbwlng Do you have insights on how this can be validated on the JSON type?